crates/sui-faucet/src/server.rs

```
@@ -35,12 +35,203 @@ use tracing::{info, warn};
+        Self {
+            data: Mutex::new(BTreeMap::new()),
+            reset_time_interval_secs,
+            max_requests_per_ip,
+        }
+    }
+
+    /// Get the current timestamp in seconds
+    fn current_timestamp_secs() -> u64 {
+        SystemTime::now()
+            .duration_since(UNIX_EPOCH)
+            .expect("Time went backwards")
+            .as_secs()
+    }
+
+    /// Checks if the user-agent is present, and if it is the expected discord bot user-agent. This
+    /// includes a unique password that the bot has set in the user-agent header.
+    fn is_discord_bot(&self, user_agent: Option<&HeaderValue>) -> Result<bool, FaucetError> {
+        if let (Some(discord_bot), Some(v)) = (DISCORD_BOT.as_ref(), user_agent) {
+            let header = v
+                .to_str()
+                .map_err(|e| FaucetError::InvalidUserAgent(e.to_string()))?;
+            Ok(header == format!("discord-bot-{}", discord_bot))
+        } else {
+            Ok(false)
+        }
+    }
+
+    /// Validates a token
+    /// - against Cloudflare turnstile's server to ensure token was issued by turnstile
+    /// - against the IP address' request count
+    async fn validate_token(
+        &self,
+        url: &str,
+        addr: SocketAddr,
+        token: &str,
+    ) -> Result<(), (StatusCode, FaucetError)> {
+        let turnstile_key = TURNSTILE_SECRET_KEY.as_ref().unwrap().as_str();
+        let req = reqwest::Client::new();
+        let params = [
+            ("secret", turnstile_key),
+            ("response", token),
+            ("remoteip", &addr.ip().to_string()),
+        ];
+
+        // Make the POST request
+        let response = req.post(url).form(&params).send().await;
+
+        match response {
+            Ok(resp) => {
+                if resp.status().is_success() {
+                    let body = resp.json::<TokenVerification>().await;
+                    if let Ok(body) = body {
```

This code seems to ignore the response from cloudflare (and continue with the rest of the function) if it cannot be parsed as a `TokenVerification` error -- is that intended? It seems wrong -- (i.e. this should fail closed, ideally).

---

Fixed!